File: Demo/Day19_OC(RunLoop)/Day19_OC(RunLoop)/ViewController.m
//
//  ViewController.m
//  Day19_OC(RunLoop)
//
//  Created by 亿存 on 2020/8/14.
//  Copyright © 2020 亿存. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@property (nonatomic , strong) NSThread *thread;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}


@end

//MARK: - 什么事RunLoop
//Run表示运行，Loop循环。结合在一起就是运行的循环的意思。通俗的来说就是「跑圈」

//MARK: - RunLoop和现成
//RunLoop和现成是息息相关额。我们知道线程的作用就是用来执行特定的一个或多个任务。在默认情况下，线程执行完就会退出，就不能再执行任务了，这时我们就需要采用一种方式来让线程能够不断的处理任务而不会退出。这就是RunLoop。
// 1. 一个线程对应一个RunLoop对象。每个线程都有唯一一个与之对应的RunLoop对象。
// 2. RunLoop并不能保证线程的安全。稚嫩恶搞在当前线程呢你不操作当前线程的RunLoop对象，而不能再当前线程内部去操作其它线程的RunLoop对象方法。
// 3. RunLoop对象在第一次获取RunLoop时创建，销毁则是在线程结束的时候。
// 4. 主线程的RunLoop对象系统自动帮我门创建好了，而子线程的RunLoop对象则需要我们主动创建和维护

File: Demo/Day9_Protocol/Day9_Protocol/ViewController.swift
//
//  ViewController.swift
//  Day9_Protocol
//
//  Created by 亿存 on 2020/7/16.
//  Copyright © 2020 亿存. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        
        //协议的定义
        //参考 ProtocolBase.swift
        
        
        //调用  作为类型使用的协议
        let randomNum = BoomTest.init(des: "💥", random: RandomNumberSmallGenerator())
        for _ in 0...2 {
            let result = randomNum.toRandom()
            print(result)
        }
        
        //MARK: - 协议类型的集合： 协议作为集合的元素类型举例：
        let protocolArray : [RandomNumberProtocol] = [RandomNumberSmallGenerator(),RandomNumberBigGenerator()]
//        for item in protocolArray {
//            let num = item.random()
//            print(num) //!< 1 17
//        }
        for item in protocolArray {
            let num = item.random()
            if let smallGenerator = item as? RandomNumberSmallGenerator {
                print(smallGenerator.name) //log:遵守协议的属性
            }
            print(num) //!< 1 17
        }
        
        //调用 有条件地遵守协议
        let num1 = RandomNumberSmallGenerator()
        let num2 = RandomNumberSmallGenerator()
        let protocolArray2 = [num1,num2]
        print("\(protocolArray2.random())")//!< 732
        
        //调用  协议组合
        let cat = Kid.hasCat(pet: Cat.init())//!< 恭喜你获得了宠物猫:小黄 颜色：黑色 特点：撒娇
        print(cat)
        let dog = Kid.hasDog(pet: Husky.init("哈怂奇", "火红", "家中地雷"))//!< 恭喜你获得了宠物狗:阿里克 颜色：火红 特点：家中地雷
        print(dog)
        
        // 检查类型实例是否遵守特定协议
        let objects: [AnyObject] = [
            Circle(radius: 2.0),
            Country(area: 243_610),
            Husky.init("哈士奇", "火红", "家中地雷")
        ]
        for object in objects {
            if let objectWithArea = object as? HasArea {
                print("Area is \(objectWithArea.area)")
            } else {
                print("Something that doesn't have an area")
            }
        }
        
        //可选协议要求
        let counter = Counter()
        counter.dataSource = ThreeSource()
        for _ in 1...4 {
            counter.increment()
            print(counter.count) // 3 6 9 12
        }
    }


}

